pydantic/_internal/_decorators.py

```diff
@@ -218,3 +226,107 @@ def in_ipython() -> bool:
         return False
     else:  # pragma: no cover
         return True
+
+
+AnyCallable = Callable[..., Any]
+
+
+def make_generic_validator(validator: AnyCallable) -> 'ValidatorFunction':  # noqa: C901
+    """
+    Make a generic function which calls a validator with the right arguments.
+
+    Unfortunately other approaches (eg. return a partial of a function that builds the arguments) is slow,
+    hence this laborious way of doing things.
+
+    It's done like this so validators don't all need **kwargs in their signature, eg. any combination of
+    the arguments "values", "fields" and/or "config" are permitted.
+    """
+    accepts_variable_kwargs = False
+    value_skipped = False
+    kwargs: Set[str] = set()
+    for parameter_name, parameter in inspect.signature(validator).parameters.items():
+        if not value_skipped:
+            if parameter.kind not in (Parameter.POSITIONAL_ONLY, Parameter.POSITIONAL_OR_KEYWORD):
+                raise TypeError  # TODO: Error message
+            value_skipped = True
+            continue
+        if parameter.kind == Parameter.VAR_KEYWORD:
+            accepts_variable_kwargs = True
+            break
+        elif parameter.kind == Parameter.VAR_POSITIONAL:
+            raise TypeError  # TODO: Error message
+        if parameter_name in {'values', 'config', 'context'}:
+            kwargs.add(parameter_name)
+        else:
+            raise TypeError  # TODO: Error message
+    if accepts_variable_kwargs:
+
+        def wrapper(
+            __input_value: Any, *, data: Any, config: CoreConfig | None, context: Any, **future_kwargs: Any
+        ) -> Any:
+            return validator(__input_value, values=data, config=config, context=context, **future_kwargs)
+
+        return wrapper
+    elif not kwargs:
+
+        def wrapper(
+            __input_value: Any, *, data: Any, config: CoreConfig | None, context: Any, **future_kwargs: Any
+        ) -> Any:
+            return validator(__input_value)
+
+        return wrapper
+    elif kwargs == {'values'}:
+
+        def wrapper(
+            __input_value: Any, *, data: Any, config: CoreConfig | None, context: Any, **future_kwargs: Any
+        ) -> Any:
+            return validator(__input_value, values=data)
+
+        return wrapper
+    elif kwargs == {'context'}:
+
+        def wrapper(
+            __input_value: Any, *, data: Any, config: CoreConfig | None, context: Any, **future_kwargs: Any
+        ) -> Any:
+            return validator(__input_value, context=context)
+
+    elif kwargs == {'config'}:
+
+        def wrapper(
+            __input_value: Any, *, data: Any, config: CoreConfig | None, context: Any, **future_kwargs: Any
+        ) -> Any:
+            return validator(__input_value, config=config)
+
+        return wrapper
```

Comment:
Do we want to raise an error for `config`? Or should we name this argument something else (`core_config`?) since it is not the same object as V1? Not sure how we handled this breaking change elsewhere with `Config` (e.g. `model.model_config`)?

Reply:
Good question.

I am leaning toward probably a good idea to change it, just to reduce confusion during migration.

Another thing to note — we should probably ensure that the typing is set up in a way that the validator decorator will error if the item whose signature it is wrapping has the wrong signature. I think this should be possible by annotating the type of the input to `Validator` with a [callback protocol](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols).

Will it be possible to maintain (deprecated) backwards compatibility with validators that currently define the config argument? If so, I think that would be an even stronger argument for giving it a different name here (and `core_config` doesn't seem unreasonable to me, though `cfg` might be nicer for reducing typing without losing clarity).

Reply:
> Will it be possible to maintain (deprecated) backwards compatibility with validators that currently define the config argument?

yes in the sense that things won't crash calling the function. no in the sense that if they used the `config` argument at all things will fail since it's a different object. so yeah I'm leaning towards not making the signature backwards compatible.

We wouldn't add a new `cfg` parameter, we'd say use `info`.

Reply:
> Or should we name this argument something else (core_config?)

If we are considering renaming to `core_config`, I'm just proposing that `cfg` might be a reasonable (and less verbose) alternative. If we'd say 'use `info`' instead of having a `cfg` argument, I'd probably suggest we error instead of changing to `core_config` as an argument.

Don't have a strong opinion though.

Reply:
I went with explicitly erroring if `config` is an acceptable keyword argument.